numpy/random/_generator.pyx

```diff
@@ -694,21 +694,23 @@ cdef class Generator:
         cdef uint64_t set_size, mask
         cdef uint64_t[::1] hash_set
         # Format and Verify input
+        a_original = a
         a = np.array(a, copy=False)
         if a.ndim == 0:
             try:
                 # __index__ must return an integer by python rules.
                 pop_size = operator.index(a.item())
             except TypeError:
-                raise ValueError("a must an array or an integer")
+                raise ValueError("a must be an array or an integer, "
+                                 f"not {type(a_original)}") from None
```

Comment:
I'm not convinced the `from None` is a good idea. Consider:
```python
class MyObject:
    def __index__(self):
        if some_complex_condition:
            raise TypeError("with useful information")
        return 42
```
or
```python
class MyObject:
    def __index__(self):
        return 1 + "2"  # oops, TypeError
```
With this PR, the traceback of `choice(MyObject())` no longer shows me error information. I think `from None` should be used _very_ sparingly (eg, for `KeyError` where you really know exactly what the cause was).

Reply:
Do you recommend using `except TypeError as exc` and then `raise ValueError(...) from exc`, or just leave it as it is in master?

Reply:
`from exc` sounds good to me.

Reply:
Done.